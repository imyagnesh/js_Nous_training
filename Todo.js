import React, { Component, createRef } from 'react';
import './todo.css';

class Todo extends Component {
  inputRef = createRef();

  state = {
    todoList: [],
    filterType: 'all',
  };

  addTodo = event => {
    event.preventDefault();
    this.setState(
      ({ todoList }) => ({
        todoList: [
          ...todoList,
          {
            id: new Date().valueOf(),
            text: this.inputRef.current.value,
            isDone: false,
          },
        ],
        filterType: 'all',
      }),
      () => {
        this.inputRef.current.value = '';
      },
    );
  };

  completeTodo = todo => {
    const { todoList } = this.state;
    const index = todoList.findIndex(x => x.id === todo.id);
    const updatedTodo = [
      ...todoList.slice(0, index),
      { ...todo, isDone: !todo.isDone },
      ...todoList.slice(index + 1),
    ];
    this.setState({
      todoList: updatedTodo,
    });
  };

  deleteTodo = todo => {
    const { todoList } = this.state;
    const index = todoList.findIndex(x => x.id === todo.id);
    const updatedTodo = [
      ...todoList.slice(0, index),
      ...todoList.slice(index + 1),
    ];
    this.setState({
      todoList: updatedTodo,
    });
  };

  filterTodo = filterType => {
    this.setState({
      filterType,
    });
  };

  filteredList = () => {
    const { todoList, filterType } = this.state;
    return todoList.filter(todoItem => {
      switch (filterType) {
        case 'completed':
          return todoItem.isDone;

        case 'pending':
          return !todoItem.isDone;

        default:
          return true;
      }
    });
  };

  render() {
    return (
      <div className="todo-wrapper">
        <h1>Todo App</h1>

        <form onSubmit={this.addTodo}>
          <input type="text" name="todo" ref={this.inputRef} />
          <button type="submit">Add Todo</button>
        </form>
        <div className="list-wrapper">
          {this.filteredList().map(todoItem => (
            <div className="list-container" key={todoItem.id}>
              <input
                type="checkbox"
                name="isDone"
                id="isDone"
                checked={todoItem.isDone}
                onChange={() => this.completeTodo(todoItem)}
              />
              <span
                style={{
                  textDecoration: todoItem.isDone ? 'line-through' : 'none',
                }}>
                {todoItem.text}
              </span>
              <button type="button" onClick={() => this.deleteTodo(todoItem)}>
                Delete
              </button>
            </div>
          ))}
        </div>
        <div className="filter-wrapper">
          <button type="button" onClick={() => this.filterTodo('all')}>
            All
          </button>
          <button type="button" onClick={() => this.filterTodo('pending')}>
            Pending
          </button>
          <button type="button" onClick={() => this.filterTodo('completed')}>
            Completed
          </button>
        </div>
      </div>
    );
  }
}

export default Todo;
