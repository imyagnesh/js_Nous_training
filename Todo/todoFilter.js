import React, { memo } from 'react';
import PropTypes from 'prop-types';

const TodoFilter = ({ filterTodo }) => {
  console.log('TodoFilter render');
  return (
    <div data-testid="filter-wrapper" className="filter-wrapper">
      <button type="button" onClick={val => filterTodo('all')}>
        All
      </button>
      <button type="button" onClick={() => filterTodo('pending')}>
        Pending
      </button>
      <button type="button" onClick={() => filterTodo('completed')}>
        Completed
      </button>
    </div>
  );
};

TodoFilter.propTypes = {
  filterTodo: PropTypes.func.isRequired,
};

export default memo(TodoFilter);
