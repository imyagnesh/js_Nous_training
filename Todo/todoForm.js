import React, { forwardRef, memo } from 'react';
import PropTypes from 'prop-types';

const TodoForm = forwardRef(({ addTodo, status }, ref) => {
  console.log('TodoForm render');
  return (
    <form onSubmit={addTodo}>
      <input type="text" name="todo" ref={ref} />
      <button
        type="submit"
        disabled={status.some(x => x.type === 'add_todo_request')}>
        Add Todo
      </button>
    </form>
  );
});

TodoForm.propTypes = {
  addTodo: PropTypes.func.isRequired,
};

export default memo(TodoForm);
