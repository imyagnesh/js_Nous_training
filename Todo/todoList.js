import React, { memo } from 'react';
import PropTypes from 'prop-types';

const TodoList = ({ filteredList, completeTodo, deleteTodo, status }) => {
  console.log('TodoList render');
  return (
    <div className="list-wrapper">
      {status.some(x => x.type === 'fetch_todos_request') ? (
        <h2>Loading....</h2>
      ) : (
        <>
          {filteredList.map(todoItem => (
            <div className="list-container" key={todoItem.id}>
              <input
                type="checkbox"
                name="isDone"
                id="isDone"
                checked={todoItem.isDone}
                disabled={status.some(
                  x => x.type === 'update_todo_request' && x.id === todoItem.id,
                )}
                onChange={() => completeTodo(todoItem)}
              />
              <span
                style={{
                  textDecoration: todoItem.isDone ? 'line-through' : 'none',
                }}>
                {todoItem.text}
              </span>
              <button
                type="button"
                disabled={status.some(
                  x => x.type === 'delete_todo_request' && x.id === todoItem.id,
                )}
                onClick={() => deleteTodo(todoItem)}>
                Delete
              </button>
            </div>
          ))}
        </>
      )}
    </div>
  );
};

TodoList.propTypes = {
  filteredList: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      text: PropTypes.string,
      isDone: PropTypes.bool,
    }),
  ).isRequired,
  completeTodo: PropTypes.func.isRequired,
  deleteTodo: PropTypes.func.isRequired,
};

export default memo(TodoList);
