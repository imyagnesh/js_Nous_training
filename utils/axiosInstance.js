import axios from 'axios';

const axiosInstance = axios.create({
  baseURL: 'http://localhost:3000',
  timeout: 1000,
  timeoutErrorMessage: 'Time Out. Plase try again',
});

axiosInstance.interceptors.request.use(
  function (config) {
    // Do something before request is sent
    const token = sessionStorage.getItem('token');
    if (token) {
      config.headers = {
        ...config.headers,
        Authorization: token,
      };
    }
    return config;
  },
  function (error) {
    // Do something with request error
    return Promise.reject(error);
  },
);

// Add a response interceptor
axiosInstance.interceptors.response.use(
  function (response) {
    // Any status code that lie within the range of 2xx cause this function to trigger
    // Do something with response data
    return response;
  },
  function (error) {
    // Any status codes that falls outside the range of 2xx cause this function to trigger
    // Do something with response error
    return Promise.reject(error);
  },
);

export default axiosInstance;
