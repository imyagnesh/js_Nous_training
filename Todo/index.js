import React, { Component, createRef } from 'react';
import './todo.css';
import TodoFilter from './todoFilter';
import TodoForm from './todoForm';
import TodoList from './todoList';
import axiosInstance from '@utils/axiosInstance';

class Todo extends Component {
  inputRef = createRef();

  state = {
    todoList: [],
    filterType: 'all',
    status: [],
  };

  componentDidMount() {
    this.fetchTodos();
  }

  changeStatus = (statusType, id) => {
    const [type, requestName] = /(.*)_(request|fail)/.exec(statusType);
    this.setState(({ status }) => {
      const index = status.findIndex(x => {
        if (id) {
          return x.type.includes(requestName) && x.id === id;
        }
        return x.type.includes(requestName);
      });
      if (index === -1) {
        return {
          status: [...status, { type, id }],
        };
      }
      return {
        status: [
          ...status.slice(0, index),
          { type, id },
          ...status.slice(index + 1),
        ],
      };
    });
  };

  fetchTodos = async filterType => {
    let params = {};
    if (
      filterType &&
      (filterType === 'pending' || filterType === 'completed')
    ) {
      params.isDone = filterType === 'completed';
    }
    try {
      this.changeStatus('fetch_todos_request');
      const res = await axiosInstance.get(`todoList`, {
        params,
      });
      this.setState(({ status }) => ({
        todoList: res.data,
        filterType: filterType || 'all',
        status: status.filter(x => !x.type.includes('fetch_todos')),
      }));
    } catch (error) {
      this.changeStatus('fetch_todos_fail');
    }
  };

  addTodo = async event => {
    try {
      this.changeStatus('add_todo_request');
      event.preventDefault();
      const res = await axiosInstance.post('todoList', {
        text: this.inputRef.current.value,
        isDone: false,
      });

      this.setState(
        ({ todoList, status }) => ({
          todoList: [...todoList, res.data],
          filterType: 'all',
          status: status.filter(x => !x.type.includes('add_todo')),
        }),
        () => {
          this.inputRef.current.value = '';
        },
      );
    } catch (error) {
      this.changeStatus('add_todo_fail');
    }
  };

  completeTodo = async todo => {
    try {
      this.changeStatus('update_todo_request', todo.id);
      const res = await axiosInstance.put(`todoList/${todo.id}`, {
        ...todo,
        isDone: !todo.isDone,
      });

      const { todoList } = this.state;
      const index = todoList.findIndex(x => x.id === todo.id);
      const updatedTodo = [
        ...todoList.slice(0, index),
        res.data,
        ...todoList.slice(index + 1),
      ];
      this.setState(({ status }) => ({
        todoList: updatedTodo,
        status: status.filter(x => x.id !== todo.id),
      }));
    } catch (error) {
      this.changeStatus('update_todo_fail', todo.id);
    }
  };

  deleteTodo = async todo => {
    try {
      this.changeStatus('delete_todo_request', todo.id);

      await axiosInstance.delete(`todoList/${todo.id}`);

      const { todoList } = this.state;
      const index = todoList.findIndex(x => x.id === todo.id);
      const updatedTodo = [
        ...todoList.slice(0, index),
        ...todoList.slice(index + 1),
      ];
      this.setState(({ status }) => ({
        todoList: updatedTodo,
        status: status.filter(x => x.id !== todo.id),
      }));
    } catch (error) {
      this.changeStatus('delete_todo_fail', todo.id);
    }
  };

  render() {
    const { todoList, status } = this.state;

    console.log(status);

    return (
      <div className="todo-wrapper">
        <h1>Todo App</h1>

        <TodoForm addTodo={this.addTodo} ref={this.inputRef} status={status} />

        <TodoList
          filteredList={todoList}
          completeTodo={this.completeTodo}
          deleteTodo={this.deleteTodo}
          status={status}
        />

        <TodoFilter filterTodo={filterType => this.fetchTodos(filterType)} />
      </div>
    );
  }
}

export default Todo;
