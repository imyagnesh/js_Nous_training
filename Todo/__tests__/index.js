import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import MockAdapter from 'axios-mock-adapter';
import Todo from '../index';
import axiosInstance from '@utils/axiosInstance';

const mock = new MockAdapter(axiosInstance);

describe('Test todo application', () => {
  test('should check component render without error', () => {
    const { container } = render(<Todo />);
    expect(container.firstChild).toBeTruthy();
  });

  test('should render data on component load', async () => {
    mock.onGet('/todoList').reply(200, [
      {
        id: 1,
        text: 'Get Milk',
        isDone: true,
      },
    ]);

    const { container } = render(<Todo />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
