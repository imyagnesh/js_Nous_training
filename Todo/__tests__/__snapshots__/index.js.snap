// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test todo application should render data on component load 1`] = `
<div
  class="todo-wrapper"
>
  <h1>
    Todo App
  </h1>
  <form>
    <input
      name="todo"
      type="text"
    />
    <button
      type="submit"
    >
      Add Todo
    </button>
  </form>
  <div
    class="list-wrapper"
  >
    <h2>
      Loading....
    </h2>
  </div>
  <div
    class="filter-wrapper"
    data-testid="filter-wrapper"
  >
    <button
      type="button"
    >
      All
    </button>
    <button
      type="button"
    >
      Pending
    </button>
    <button
      type="button"
    >
      Completed
    </button>
  </div>
</div>
`;

exports[`Test todo application should take snapshot 1`] = `
<div
  class="todo-wrapper"
>
  <h1>
    Todo App
  </h1>
  <form>
    <input
      name="todo"
      type="text"
    />
    <button
      type="submit"
    >
      Add Todo
    </button>
  </form>
  <div
    class="list-wrapper"
  >
    <h2>
      Loading....
    </h2>
  </div>
  <div
    class="filter-wrapper"
    data-testid="filter-wrapper"
  >
    <button
      type="button"
    >
      All
    </button>
    <button
      type="button"
    >
      Pending
    </button>
    <button
      type="button"
    >
      Completed
    </button>
  </div>
</div>
`;
